keps/sig-api-machinery/4420-retry-generate-name/README.md

```diff
@@ -0,0 +1,550 @@
+  - Verify that a resource is generated with the max retries when conflicting names
+    exist
+  - Verify that if max retries is exceeded that a already exists error is returned
+  - Verify that retries are not attempted when feature is disabled
+
+##### Integration tests
+
+<!--
+Integration tests are contained in k8s.io/kubernetes/test/integration.
+Integration tests allow control of the configuration parameters used to start the binaries under test.
+This is different from e2e tests which do not allow configuration of parameters.
+Doing this allows testing non-default options and multiple different and potentially conflicting command line options.
+-->
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+-->
+
+- Verify that retries are not attempted when feature is disabled
+- Verify that up to max retries are attempted when conflicting names exist
+- Verify that if max retries is exceeded that a already exists error is returned
+
+##### e2e tests
+
+This feature is not e2e testable. There is no predictable way to
+trigger name collisions.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Unit tests added. The `generic/registry/store_test.go` suite covers this functionality remarkably well.
+
+#### Beta
+
+- Gather feedback from developers and surveys
+- Add integration tests
+- Additional tests are in Testgrid and linked in KEP
+- Benchmark worst case: Max retries are attempted for all create requests.
+- Consider further optimizing. One possible option:  If the
+  first create attempt conflicts, do a loop to check etcd
+  for all subsequent name generations to make sure the name
+  doesn't exist before retrying the create request.  This
+  would be more efficient for cases where multiple retries
+  are required while still keeping the non-conflicting case
+  fast.
+
+#### GA
+
+- Add `generate_name_retries: {count}` label to `apiserver_request_duration_seconds`
+  metric.
+
+### Upgrade / Downgrade Strategy
+
+N/A. This only reduces the probability of name conflicts when a request is
+served by an apiserver with this feature enabled.
+
+### Version Skew Strategy
+
+N/A. This only reduces the probability of name conflicts when a request is
+served by an apiserver with this feature enabled.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: NameGenerationRetries
+  - Components depending on the feature gate: kube-apiserver
+
+
+###### Does enabling the feature change any default behavior?
+
+No. It only reduces the probability of a error that can occur randomly.
```

Comment:
I think the answer might be “Yes”; the default is now to retry (and users can see this if they look at the relevant metrics).